cmd/sops/subcommand/publish/publish.go

```diff
@@ -0,0 +1,114 @@
+package publish
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"go.mozilla.org/sops"
+	"go.mozilla.org/sops/cmd/sops/codes"
+	"go.mozilla.org/sops/cmd/sops/common"
+	"go.mozilla.org/sops/config"
+	"go.mozilla.org/sops/keyservice"
+	"go.mozilla.org/sops/logging"
+
+	"github.com/sirupsen/logrus"
+)
+
+var log *logrus.Logger
+
+func init() {
+	log = logging.NewLogger("PUBLISH")
+}
+
+type Opts struct {
+	Interactive bool
+	ConfigPath  string
+	InputPath   string
+	KeyServices []keyservice.KeyServiceClient
+	InputStore  sops.Store
+}
+
+func Run(opts Opts) error {
+	var fileContents []byte
+	path, err := filepath.Abs(opts.InputPath)
+	if err != nil {
+		return err
+	}
+	info, err := os.Stat(path)
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		return fmt.Errorf("can't operate on a directory")
+	}
+	_, fileName := filepath.Split(path)
+
+	conf, err := config.LoadForFile(opts.ConfigPath, opts.InputPath, make(map[string]*string))
+	if err != nil {
+		return err
+	}
+	if conf.Destination == nil {
+		return errors.New("No destination configured for this file.")
+	}
+
+	// Check that this is a sops-encrypted file
+	tree, err := common.LoadEncryptedFile(opts.InputStore, opts.InputPath)
+	if err != nil {
+		return err
+	}
+
+	// Re-encrypt if settings exist to do so
+	if len(conf.ReEncryptionKeyGroups) != 0 {
+		key, err := tree.Metadata.GetDataKeyWithKeyServices(opts.KeyServices)
+		if err != nil {
+			return common.NewExitError(err, codes.CouldNotRetrieveKey)
+		}
+		tree.Metadata.KeyGroups = conf.ReEncryptionKeyGroups
+		tree.Metadata.ShamirThreshold = min(tree.Metadata.ShamirThreshold, len(tree.Metadata.KeyGroups))
+		errs := tree.Metadata.UpdateMasterKeysWithKeyServices(key, opts.KeyServices)
+		if len(errs) > 0 {
+			return fmt.Errorf("error updating one or more master keys: %s", errs)
+		}
+
+		fileContents, err = opts.InputStore.EmitEncryptedFile(*tree)
+		if err != nil {
+			return common.NewExitError(fmt.Sprintf("Could not marshal tree: %s", err), codes.ErrorDumpingTree)
+		}
+	} else {
+		fileContents, err = ioutil.ReadFile(path)
+		if err != nil {
+			return fmt.Errorf("could not read file: %s", err)
+		}
+	}
+
+	if opts.Interactive {
+		var response string
+		for response != "y" && response != "n" {
+			fmt.Printf("\nuploading %s to %s ? (y/n): ", path, conf.Destination.Path(fileName))
+			_, err := fmt.Scanln(&response)
+			if err != nil {
+				return err
+			}
+		}
+		if response == "n" {
```

Comment:
there is an infinite number of inputs people use to signify "not yes", so perhaps using a switch case with a default that cancels publishing is safer.

Reply:
For operations like this one that could potentially break things, I like terraform's approach of only accepting `yes` and nothing else as a prompt to continue, and everything else cancels the operation. Although that'd make us inconsistent with updatekeys...

Reply:
I'm happy to come back through and align `updatekeys` and the `keyservice` prompt if we determine that only `yes` works. I am a fan of that, or making `no` the default if you just press enter.

Reply:
I'm not 100% that breaking current behavior for this change is worth it. I doubt there's anyone doing `yes | sops ...` but who knows...